# Term Frequencies for talks by gender

In previous notebooks, we have used speakers' genders to apply a gender label to each talk. The gender labels used are: male, female, and not one gender. We note that the list of speakers in the corpus of TED talks identify on the gender binary, at least in the public facing documents about them (which include their TED talk bios and third party websites and publicity). 

## Summary

In this notebook, we will create document-term-frequencies for each set of gendered talks. This notebook follows closely the term-frequency notebook for the full corpus (`01-Terms.ipynb`). We will take each subset separately, annotating the full procedure for just for one of the three subsets. 

Similarly to `01-Terms.ipynb`, the primary task of this notebook is to count the words in the texts of each gender subset. To do so, we also must address the matter of parentheticals head on. Much of this work occurs in a separate notebook, and the explanations of how that work relates to this process is detailed in `01-Terms.ipynb`.

In this notebook, we will take each subset individually (talks labeled male, female, and not one gender) and complete the following procedure: 
* Remove the parentheticals
* Create term matrix from the resulting vocabulary
* Adjust the threshold for a word to be included in the term-matrix
* Export the resulting as a CSV


### Loading necessary packages and functions


In [2]:
# IMPORTS

import pandas as pd, re, csv, nltk
from sklearn.feature_extraction.text import CountVectorizer

# LOCAL FUNCTION --> Explained in 01-Terms.ipynb

def remove_parentheticals(text):
    new_text = text
    for rgx_match in parentheticals:
        new_text = re.sub(rgx_match, ' ', new_text.lower(), flags=re.IGNORECASE)
    return new_text

parentheticals = [ "\(laughter\)", "\(applause\)", "\(music\)", "\(video\)", 
                  "\(laughs\)", "\(applause ends\)", "\(audio\)", "\(singing\)", 
                  "\(music ends\)", "\(cheers\)", "\(cheering\)", "\(recording\)", 
                  "\(beatboxing\)", "\(audience\)", "\(guitar strum\)", 
                  "\(clicks metronome\)", "\(sighs\)", "\(guitar\)", "\(marimba sounds\)", 
                  "\(drum sounds\)" ]

## Talks labeled Female

In this section, we will expand on the four step laid out in the summary above for the talks labeled as "female." This labelling was applied to talks where all the speakers identified with female pronouns. 

We begin by loading the female-labeled talks and taking a peak at the first 5 rows: 

In [3]:
# DATA
talks_female = pd.read_csv('talks_female.csv', index_col='Talk_ID') # Skipping `index_col='Talk_ID'`
talks_female.shape

(260, 14)

In [4]:
talks_female.head()

public_url  \
Talk_ID                                                      
53       https://www.ted.com/talks/majora_carter_s_tale...   
86       https://www.ted.com/talks/julia_sweeney_on_let...   
55       https://www.ted.com/talks/jehane_noujaim_inspi...   
46       https://www.ted.com/talks/jennifer_lin_improvs...   
45       https://www.ted.com/talks/sirena_huang_dazzles...   

                                headline  \
Talk_ID                                    
53                   Greening the ghetto   
86                     Letting go of God   
55         My wish: A global day of film   
46         Improvising on piano, aged 14   
45       An 11-year-old's magical violin   

                                               description    event duration  \
Talk_ID                                                                        
53       In an emotionally charged talk, MacArthur-winn...  TED2006  0:18:36   
86       When two young Mormon missionaries knock on Ju...  TED2006  0:16:32   
55       Jehane Noujaim unveils her 2006 TED Prize wish...  TED2006  0:25:38   
46       Pianist and composer Jennifer Lin gives a magi...  TED2004  0:24:05   
45       Violinist Sirena Huang gives a technically bri...  TED2006  0:24:41   

        published                                               tags    views  \
Talk_ID                                                                         
53        6/27/06  MacArthur grant,cities,green,activism,politics...  2000421   
86        7/10/06  atheism,Christianity,religion,God,comedy,humor...  3903747   
55        7/25/06  TED Prize,peace,entertainment,movies,global is...   409921   
46         8/8/06  wunderkind,entertainment,piano,creativity,musi...  1680398   
45         8/8/06  violin,wunderkind,entertainment,youth,music,pe...  2811646   

                                                      text       speaker_1  \
Talk_ID                                                                      
53         If you're here today — and I'm very happy th...   Majora Carter   
86         On September 10, the morning of my seventh b...   Julia Sweeney   
55         I can't help but this wish: to think about w...  Jehane Noujaim   
46         (Music)    (Music ends)    (Applause)    Tha...    Jennifer Lin   
45         (Music)    (Music ends)    (Applause)    (Ap...    Sirena Huang   

        speaker_2  speaker_3  speaker_4 talk_gender  
Talk_ID                                              
53            NaN        NaN        NaN      female  
86            NaN        NaN        NaN      female  
55            NaN        NaN        NaN      female  
46            NaN        NaN        NaN      female  
45            NaN        NaN        NaN      female

## Raw Frequencies

In this notebook, we are only concerned with the texts of each TED talk. Therefore, we begin by creating a new data object that only has the texts of the talks themselves. To uniquely identify each talk, we will also create a second data object that lists the talk IDs in the same order as the talks' texts. The talk ID is unique for each talk and will allow for results from this notebook to be tied to the metadata about each talk (which are stored in another file). 

In [5]:
f_texts = talks_female.text.tolist()

**Nota bene**: If you are using this notebook to re-create the basic TF matrix you can stop here and jump to the end of this notebook.

There are a number of ways to get term frequencies, but **SciKit-Learn**'s `CountVectorizer` offers a one-stop shop for generating a DTM from which we can examine words as well as generate BoW-products like topic models. This is made possible by the interoperability between `CountVectorizer` and the other vectorizers and models available in `sklearn`.



In our first experiment, we run `CountVectorizer` unadorned. The default options are: lowercase everything, get rid of all punctuation, make a word out of anything more than two characters long. The only thing that might not be welcome is the splitting of contractions. For now, we will leave things as they are. (Also, please note, no stopwords were used, so we have an unfiltered word list and no pre-processing of the texts is done.)

CHECK THIS ASSERTION: Since `CountVectorizer` only counts when *fitting* and does not engage in any kind of normalization or centering, *transforming*, there is no need to break the two functionalities out in the code that follows.

In [7]:
# If you want to pass options, pass them here:
vectorizer = CountVectorizer( lowercase = True )

# fit the model to the data 
# vecs = vectorizer.fit(texts)
X = vectorizer.fit_transform(f_texts)

# see how many features we have
X.shape

(260, 20990)

In [8]:
X

<260x20990 sparse matrix of type '<class 'numpy.int64'>'
	with 148171 stored elements in Compressed Sparse Row format>

20,990 tokens is our raw, unfiltered, no pre-processing baseline. It includes, as we will see, a number of artifacts of the TED transcription process, including a variety of ways to spell out *ah*, the use of numerals for a host of things -- from dates to counts and everything in between, and some things which actually take places outside the talk itself, like audience laughter, which we here term *parentheticals*. (Accounting for all these artifacts requires a number of closer inspections, which take place in a separate notebook: [Term Artifacts](Terms-02).)

In [9]:
raw_df = pd.DataFrame(X.toarray(), columns = vectorizer.get_feature_names())
raw_series = raw_df.sum()

In [35]:
# So at least our first 340 "words" are numbers

raw_df.iloc[1:5,345]

1    0
2    0
3    0
4    0
Name: a10, dtype: int64

In [12]:
raw_series.head(40)

00             3
000          224
000000004      1
06             1
07             1
080            1
10           173
100           93
1000th         1
100th          1
101            3
102            1
106            1
10b            1
10th           6
11            52
111            1
114            1
115            1
118            2
11th           6
12            74
120            7
1214b          1
124            1
125            1
128            1
129            1
12th           1
13            26
130            7
1313           1
134            1
135            1
137            5
13th           1
14            34
140           10
14th           1
15            84
dtype: int64

A hand inspection of the output above turned up a couple of interesting issues: that there are parentheticals mixed in with the text of the talks (see 01-Terms-02-Parentheticals) and numbers feature in the talks regularly (see 01-Terms-03-Numbers). 

There are also some interesting tokenization dimensions to a crowd-sourced transcription technology:

| Term    | Freq |
|:---------|---:|
|aa       |12 |
|aaa      |7  |
|aaaa     |2  |
|aaaaa    |1  |
|aaaaaaaah|1  |
|aaaaaaah |1  |
|aaaaaah  |3  |
|aaaah    |2  |
|aaaahhh  |1  |
|aaah     |3  |
|aag      |1  |
|aah      |10 |

There are an equal number of alternate spellings for *shh*. There's also this oddity:

| Term    | Freq |
|:---------|---:|
|FALSE    |115|
|TRUE     |909|

It appears to be simply the occurrences of *true* and *false* but somehow the vectorizer thinks they are Booleans?

The rest of the hand-examination turns up no other issues, so the basic vectorization built into `sklearn` appears to be satisfactory, with the only exception being its breaking of contractions at the apostrophe: e.g., *isn't* becomes *isn* and *'t*, with the latter being thrown away as too small -- which means that the indefinite article *a(n)* is also not present in the frequencies, but the definite article *the* remains. 

## Revised Frequencies without Parentheticals

For more on the work leading up to this regex, see the `01-Terms-02-Parentheticals` notebook: it contains a number of experiments, which were vexed by sklearn's vectorizer only accepting strings as inputs. (While gensim appears to accept texts as lists of words, we are hoping, I think, to keep our code base as simple as possible.) The eventual workaround is to clean the texts of the top 20 parentheticals, joining the list back to a string, and then pushing the result to sklearn. This is less than optimum, but the hack works and appears to be a hack performed elsewhere by others.

For now, the first thing we have is the list of the top 20 parentheticals, some of which are two words:

In [30]:
noparens_vec = CountVectorizer(preprocessor = remove_parentheticals)
noparens_X = noparens_vec.fit_transform(texts)
noparens_X.shape

(260, 20988)

Good news. Our removal parentheticals has only removed two words. 

In [31]:
noparens_df = pd.DataFrame(noparens_X.toarray(), columns = noparens_vec.get_feature_names())

In [32]:
term = 'laughter'
f"The baseline count for *{term}* is {raw_df[term].sum()}; \
with cleaning it is {noparens_df[term].sum()}"

'The baseline count for *laughter* is 1071; with cleaning it is 66'

## Frequencies of Terms with a 2-Document Minimum

Setting a `min_df = 2` removes all words that appear in only document. (An exploration of *one document wonders* is elsewhere.) As we can see from the `shape` below, we dropped 50379 to 29340, resulting in a reduction of 42% of the possible features.

In [33]:
min2_vec = CountVectorizer(preprocessor = remove_parentheticals, min_df = 2)
min2_X = min2_vec.fit_transform(texts)
min2_X.shape

(260, 10805)

As above we can sum our terms and write the results to a CSV file for closer inspection. Here, we highlight just a particular series of words.

In [34]:
# Create the dataframe
min2_df = pd.DataFrame(min2_X.toarray(), columns = min2_vec.get_feature_names())

# Create a series of sums
min2_series = min2_df.sum()

# Write sums to CSV:
# min2_series.to_csv('../output/word_freq_clean.csv')

# Inspect the sums
min2_series[1000:1005]

batteries     2
battery       9
battle       13
battles       5
bay           9
dtype: int64

## Frequencies of Terms Appearing in All Documents

Having trimmed terms that occur in only document from our feature set, we also need to determine what words occur across all the documents and thus are not particularly effective discriminators of topic. 

Note that this is a somewhat more dangerous move to make, since it has been shown that one of the keys to genre and gender in documents are the varying uses of so-called function words. In fact, as our explorations establish, using sklearn's built-in algorithms reveals a focus on removing common words does not really result in much of a reduction in the overall feature set, only 8 words occurring across 99% or more documents and that number only rising to 46 when we change the minimum threshold to 90%.

In [35]:
alldocs_vec = CountVectorizer(preprocessor = remove_parentheticals, min_df = 0.99)
alldocs_X = alldocs_vec.fit_transform(texts)
alldocs_X.shape

(260, 7)

In [36]:
alldocs_df = pd.DataFrame(alldocs_X.toarray(), columns = alldocs_vec.get_feature_names())
alldocs_series = alldocs_df.sum()
alldocs_series

and     17437
in       8529
of      12139
the     21585
this     5038
to      14453
you      7777
dtype: int64

In [37]:
mostdocs_vec = CountVectorizer(preprocessor = remove_parentheticals, min_df = 0.90)
mostdocs_X = mostdocs_vec.fit_transform(texts)
mostdocs_X.shape

(260, 41)

In [42]:
# mostdocs_df = pd.DataFrame(mostdocs_X.toarray(), columns = mostdocs_vec.get_feature_names())
# mostdocs_series = mostdocs_df.sum()
# mostdocs_series.head(46)

This last run simply verifies that the exploration above with `min_df` holds true when we switch to `max_df`. That is, that our exploration of the parameters were not asking something unexpected by the code and possibly generating inexplicable and undesirable results.

In [38]:
# tnt = topsntails
tnt_vec = CountVectorizer(preprocessor = remove_parentheticals, max_df = 0.9, min_df = 2)
tnt_X = tnt_vec.fit_transform(texts)
tnt_X.shape

(260, 10766)

The numbers add up. With that done, my recommendation is that we do not, for the time being, throw away common and/or function words: that is, let's not use a stopword list. Those words could be important in other ways, and dropping them has only a limited impact on the actual document-term matrix, one that pales by comparison to the words that occur only in one document.